migration-to-python-3-only.md

```diff
@@ -0,0 +1,217 @@
+
+It is also less confusing for downstream distributor that will only see 1
+available source.
+
+
+# Notebook & Python 3 support.
+
+The Jupyter Protocol is designed to be language agnostic, so the language the
+Notebook server (or any frontend) is written in does not have to match the
+language of code within a notebook.
+
+For the next+1 version of the Notebook server (6.0), we propose to require Python 3.
+
+The rationale being that the notebook application is one of those that can
+benefit the most from recent Python improvement, such as asyncio or function annotations.
+
+Unlike for IPython, we do not believe a LTS support is necessary, as a Python 3
+only version of the notebook application can perfectly run a Python 2 kernel.
+
+In order to allow a progressive transition to Python 3 for users and downstream
+integrators, we propose to release the first Python 3 only version of the
+notebook 6 months after the first release of the Python 3 only version of IPython.
+
+We propose to clearly communicate on the next release of the notebook (5.0), that
+one of the next versions will require a Python 3 environment.
+
+Timelines:
+    - Notebook 5.0 – Python 2 compatible – Summer/Fall 2016
+    - IPython 6.0 – Python 3 Only – Early 2017
+    - Notebook 6.0 (or 7.0) – Python 3 only – Mid 2017
+
+
+## Challenges
+
+As for IPython above; the notebook package need to be able to express that it
+requires Python 3. This is mitigated by people installing `jupyter`: we can
+make that conditionally depend on different version of `notebook`, depending on
+the version of Python.
+
+In particular for Anaconda, one path forward would be to bundle the
+notebook with its own Python 3 interpreter. This can lead to some
+inconsistency for notebook server extensions, as on Python 2 Anaconda, the
+notebook server will be in a separate environment, and the notebook extensions
+and optional dependencies need to be installed in this environment as well. We
+are thus speaking of `nbconvert`, custom exporters, custom handler and server
+side extension.
+
+This is not a problem for most Linux package managers, where the norm is to
+treat Python 2 and 3 as distinct packages. In this context, the notebook will
+have a dependency on `python3`.
+
+# Not in this roadmap:
+
+
+- `nbconvert` CLI: Can likely be migrated to Python 3 only in roughly same schedule than notebook.
+- `nbformat` : suggest to keep Python 2 support as long as we support IPython 5.x, maybe later, to allow people to "upgrade" their notebook and read newer version on "old" notebook.
+- `jupyter_client` : Suggest keeping some Python 2 compatibility as this might be helpful for some downstream project.
+- `ipyparallel`: Decision is on ipyparallel users/devs.
+- `ipywidgets`
+- `nbgrader`
+- `jupyterhub` is already Py3 only
+- `jupyter_console`
+
+
+# Misc
+
+We propose to move forward with signing the "no more legacy python pledge"
+
+https://gist.github.com/brettcannon/2f6926dc6a7874478ccd
```

Comment:
> If you want to do this I would suggest taking a bit more time and input from the broader community on writing the pledge (not that there's anything wrong with this, but I get the impression that @brettcannon just cobbled that together, so it should just be considered a first draft).

Yes, I just want to make the document a bit clearer on intention and vocabulary, before pinging some Python 2 supporters on that and then having broader input from the internet.



> That being said, if you all want to make it a formal thing I fully support you, but I don't' know if I should be driving it as it might be viewed as biased coming from a Python core dev whose name is on the PEP listing the overall feature set of Python 3. ;)

I'll make it clearer that we want to likely sign something "similar to the pledge",and of course get broader feedback on it as well.